### DISCUSSION TITLE
Voiding streams with negative debt for improved accounting

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/flow#57] [Ideas]
## Context

Michael Lewellen from OpenZeppelin is interested in beta testing our OpenEnded product. [I met him in person in London](https://www.notion.so/sablier/Michael-Lewellen-OpenZeppelin-e214b4026ea1468889c4673ec2106e86), and presented our plans to him.

During our conversation about their needs, I came up with this idea.

## Idea

Add a new function called `void` with the following **requirements**:

- Can only be called by the recipient
- Can only be called when the debt exceeds zero

And the following **behavior**:

- Sets the debt to zero
- Stops the stream from accruing debt in the future

## Benefits

Implementing this feature would improve the accounting experience since it would enable recipients to signal to their accountants that "hey, we are no longer observing stream X, please ignore it".

Furthermore, this would provide Sablier with a competitive advantage over LlamaPay, which to my knowledge doesn't implement such a feature.

CC @sablier-labs/solidity and @razgraf

**Comment by razgraf** [comment]
Should we also consider having this for the sender?

Debt is a "tool" to track promised future income, so it's not locked/guaranteed, meaning senders can choose to skip it anyway. Voiding could cleanup dashboards on their side if they forget to officially close some payment stream.

Or is this contrary to what a sender can do? E.g. you either (as a sender) ask the recipient to void debt or you have to pay it up

**Reply by PaulRBerg** [reply]
I could be wrong, but I think that `cancel` and `restartStream` already enable the sender to achieve the same effect.

But then, as you said, if we explicitly allow senders to `void` streams, we break the social contract that the recipient is entitled to some debt. The recipient is a creditor, so it's their right to forgive the debt, whereas the sender is a debtor, and it's awkward to allow a debtor to explicitly void their debt (even though they can implicitly do this by choosing not to fund the stream).

Anyway, are you are on board with the idea, at least for the recipient? It's not clear from your comment.

**Reply by razgraf** [reply]
Yup, unwanted `debt` is nasty for account-level computation (e.g. aggregating how much a sender is required to top-up, or how much someone is owed) so ways to cleanup this debt would improve the UX (esp. financial overviews).

**Reply by PaulRBerg** [reply]
super

**Reply by smol-ninja** [reply]
`cancel` and `restartStream` do not set debt to zero. It sets `rps` to 0.

**Reply by andreivladbrg** [reply]
> cancel and restartStream do not set debt to zero. It sets rps to 0.

yes, and it implicitly sets the debt to zero, since in the `cancel` function no transfer will be made to the sender due to the debt

**Reply by smol-ninja** [reply]
So when they `restart` the stream, debt begins with zero?

**Reply by andreivladbrg** [reply]
yes, the debt is a dynamic amount calculated based on the balance(state variable) and dynamic streamed amount

**Comment by andreivladbrg** [comment]
> Add a new function called void with the following requirements:
Can only be called by the recipient
Can only be called when the debt exceeds zero
And the following behavior:
Sets the debt to zero
Stops the stream from accruing debt in the future

the debt is a consequence of the sender not depositing enough tokens, if the recipient calls this function means that he will simply won't care about the streamed tokens in the past

is this what we really want to achieve?

> Implementing this feature would improve the accounting experience since it would enable recipients to signal to their accountants that "hey, we are no longer observing stream X, please ignore it".

does it mean that it will also "abandon" the accounted amount as a loss?

**Reply by PaulRBerg** [reply]
> is this what we really want to achieve?

Yes. It's perfectly possible for creditors to wish to forgive someone's debt.

They might have a private arrangement in which they changed wallets, changed chains, exchanged fiat instead of crypto, etc. They can then call the `void` function to signal to their accountants that the open-ended stream is no longer active.

> does it mean that it will also "abandon" the accounted amount as a loss?

We as a protocol don't know and shouldn't care about that.

**Comment by smol-ninja** [comment]
Recipient being able to call `void` sounds like a good user experience.

But because it resets the "debt" which might be of any scale, shouldn't it be a decision made by both the parties? If recipient calls the `void` without informing the sender, this may not improve the accounting experience for the sender.

- Step 1: Recipient signals the sender through a transaction.
- Step 2: Sender confirms adds his confirmation through another transaction.
- Step 3: Debt sets to 0.

**Reply by PaulRBerg** [reply]
Requiring inputs from both parties sounds like a chore that would add friction without much benefit to either party.

- The sender may refuse to cooperate OR may not care enough about the recipient's accounting needs OR may be just lazy. Meanwhile, the recipient's accountants are at their wits' end, drowned in unclosed streams that keep accruing stale debt.
- If a sender sees a voided stream which they desperately wanted to repay, they can just look at the debt at the time of voiding, and make a one-time payment to the recipient.
- Consider the front-end development costs - it's extra effort that we should rather invest in other, more critical features.

**Comment by PaulRBerg** [comment]
Closing as it seems like we have consensus.

Opened an issue for todo tracking: https://github.com/sablier-labs/v2-open-ended/issues/61

**Comment by smol-ninja** [comment]
Me and @andreivladbrg were discussing today about this during our post audit call. We realized that the logic of 'void' is same as 'cancel'.

We will basically have two functions with different access controls doing the same thing.

What do you think if we just combine void and cancel into one function and call it pause?

What pause does is that it sets rps to zero, sets isCanceled to True and reset the debt. The only downside of this is psychological barrier to distinguish between the two but that can be solved with the UI experience. 

Thoughts?

**Reply by razgraf** [reply]
I think at this point it would be nice to have a state-machine (even a very rough one) where we can see what a stream can do. 

You're right to assume users will have a hard time distinguishing between cancel and actual pause, as until now we've been treading cancel as a "bad thing" (red=danger), and this pause-cancel shouldn't have the same visual treatment.

**Reply by PaulRBerg** [reply]
I still think that it would be helpful to have two functions because of the UX issues you mentioned, and this other reason:

When the recipient is the caller, the debt must be positive. Just like in [Lockup](https://github.com/sablier-labs/v2-core/blob/9657fbd0a77821a088b34d460dd1117e5600bb43/src/abstracts/SablierV2Lockup.sol#L161-L163), a recipient should NOT be allowed to cancel an ongoing open-ended stream.

In general in Solidity, I prefer thin over monolithic functions.

And agree with @razgraf, if you could create a state machine using Mermaid or [XState](https://xstate.js.org/) and throw it in a "Show and Tell" discussion, that would be helpful. As an example, see my walkthrough of the [Stream Status](https://github.com/sablier-labs/v2-core/discussions/543).

**Reply by smol-ninja** [reply]
[Click here](https://github.com/sablier-labs/v2-open-ended/discussions/70) to see open ended state machine diagram.
